### PR TITLE
FIR checker: warn unnecessary safe calls

### PR DESCRIPTION
**Pull request by jsjeon**
Safe call (`?.`) is desugared to `FirSafeCallExpression`. While visiting expressions of that type, we can report:
* UNNECESSARY_SAFE_CALL
* UNEXPECTED_SAFE_CALL

To report on the specific operator `?.`, positioning strategy `SAFE_ACCESS` is introduced. Also, desugared expression's source is corrected.

### CODE REVIEW

File: compiler/testData/diagnostics/tests/platformTypes/nullabilityWarnings/safeCall.fir.kt [hunk 1]
```diff
@@ -24,20 +24,20 @@ fun test() {
     // platform type with no annotation
     val platformJ = J.staticJ
 
-    platformNN?.foo()
+    platformNN<!UNNECESSARY_SAFE_CALL!>?.<!>foo()
     platformN?.foo()
-    platformJ?.foo()
+    platformJ<!UNNECESSARY_SAFE_CALL!>?.<!>foo()
```
Comment (jsjeon) [inline compiler/testData/diagnostics/tests/platformTypes/nullabilityWarnings/safeCall.fir.kt:29]: This is exactly same as https://github.com/JetBrains/kotlin/pull/4257/files#r604619673, and after digging a bit more, I think I found a root cause, but it requires a separate PR/commit I think. Will update here and that PR as well.

File: compiler/testData/diagnostics/tests/nullableTypes/safeCallOperators.fir.kt [hunk 2]
```diff
@@ -52,40 +52,40 @@ fun foo(a: A?) {
...
+        a<!UNNECESSARY_SAFE_CALL!>?.<!>ll[0][0]
+        a<!UNNECESSARY_SAFE_CALL!>?.<!>ll[0][0]++
+        a<!UNNECESSARY_SAFE_CALL!>?.<!>ll[0][0] = 1
         // No warning is reported because
         // 1. All kinds of green code with safe+call + invoke we identified fails with CCE if `a != null`, anyway
         // 2. In case of null value, the behavior is intended (no call performed)
-        a?.q()
-        a?.w++
+        a<!UNNECESSARY_SAFE_CALL!>?.<!>q()
+        a<!UNNECESSARY_SAFE_CALL, UNNECESSARY_SAFE_CALL!>?.<!>w++
```
[the remark this answers is not in the document]
Comment (punzki) [reply compiler/testData/diagnostics/tests/nullableTypes/safeCallOperators.fir.kt:67]: Well I know the 1st TODO for capturing array is already done (https://github.com/JetBrains/kotlin/commit/a2a4d94834705bf52a41dfdda9a41e2833d3151d) and it's a stale comment. I will put the 2nd (safe access receiver capture) on my own TODO list, and try to remember what to do. 🙂

**Review by punzki — APPROVED** [review]

File: compiler/testData/diagnostics/tests/platformTypes/nullabilityWarnings/safeCall.fir.kt (hunk 1 above)
Comment (jsjeon) [reply compiler/testData/diagnostics/tests/platformTypes/nullabilityWarnings/safeCall.fir.kt]: Discussed at https://github.com/JetBrains/kotlin/pull/4257#discussion_r605431455; will be fixed like https://github.com/JetBrains/kotlin/pull/4269; and on top of that, this one is gone, and now it's `// FIR_IDENTICAL`.

**Review by tgeng — APPROVED** [review]

**Review by mglukhikh — APPROVED** [review]

**Comment by mglukhikh** [comment]
Merged manually 👍

**Pull request closed without merging**